src/Playwright/Transport/WebSocketTransport.cs

```diff
@@ -0,0 +1,199 @@
+                    var buffer = WebSocket.CreateClientBuffer(DefaultBufferSize, DefaultBufferSize);
+
+                    while (!token.IsCancellationRequested && _webSocket.State == WebSocketState.Open)
+                    {
+                        WebSocketReceiveResult result;
+                        using (MemoryStream memoryStream = new())
+                        {
+                            do
+                            {
+                                result = await _webSocket.ReceiveAsync(buffer, token).ConfigureAwait(false);
+
+                                if (result.MessageType == WebSocketMessageType.Close)
+                                {
+                                    await
+                                        _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, string.Empty, _readerCancellationSource.Token).ConfigureAwait(false);
+                                }
+                                else
+                                {
+#if NETSTANDARD
+                                    await memoryStream.WriteAsync(buffer.Array, 0, result.Count, _readerCancellationSource.Token).ConfigureAwait(false);
+#else
+                                    await memoryStream.WriteAsync(new(buffer.Array, 0, result.Count), _readerCancellationSource.Token).ConfigureAwait(false);
+#endif
+                                }
+                            }
+                            while (!result.EndOfMessage);
+
+                            string output = Encoding.UTF8.GetString(memoryStream.ToArray());
+                            MessageReceived?.Invoke(this, new MessageReceivedEventArgs(output));
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Close(ex);
+            }
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (!disposing)
+            {
+                return;
+            }
+
+            _readerCancellationSource?.Dispose();
+            _webSocket?.Dispose();
+        }
+
+        private string GenerateUserAgent()
+        {
+            var architecture = RuntimeInformation.OSArchitecture;
+            var osAndVersion = RuntimeInformation.OSDescription;
+            var frameworkDescription = RuntimeInformation.FrameworkDescription;
+            var assemblyVersion = FileVersionInfo.GetVersionInfo(Assembly.GetExecutingAssembly().Location).FileVersion;
+            return $"Playwright/{assemblyVersion} {frameworkDescription} ({architecture}/{osAndVersion})";
+        }
+
+        private void SetOptions()
+        {
+            _webSocket.Options.SetRequestHeader("User-Agent", GenerateUserAgent());
+            foreach (var item in _options?.Headers ?? Array.Empty<KeyValuePair<string, string>>())
+            {
+                _webSocket.Options.SetRequestHeader(item.Key, item.Value);
+            }
+
+            var timeout = _options.Timeout;
```

Comment:
`connect({timeout})` implies that there is a budget of `timeout` that we have to establish the connection before returning from connect, has nothing to do with KeepAlive.

Reply:
Make sense have configured timeout for `CancellationTokenSource`